### PR TITLE
Framework for compatibility and upgrade checks

### PR DESCRIPTION
**Pull request by snazy**
Provides three compatibility-check approaches:
* Running older Nessie client versions against the current in-tree Nessie server
* Running the current in-tree Nessie client against older Nessie servers
* Upgrade scenarios:
  * Start Nessie using the "oldest" version, use the "oldest" Nessie client, run tests, stop server
  * Start Nessie using the "next" version, use the "next" Nessie client, run tests, stop server, and so on
  * Uses a database adapter that persists (RockdDB), so the "oldest" version starts with an emptry
    repository, each "next" version starts with the state of the "previous" version

Implementation:
* Uses JUnit 5 SPI functionality to "duplicate" test classes (one per Nessie version to be tested)
  using a JUnit engine implementation, tests are run using "standard" and well-known JUnit 5.
* Uses JUnit 5 extensions to properly populate and use current/old Nessie server/client instances and
  control their lifecycle - plus "helper" stuff like cached class loaders and such.

This change does not include required compatibility and upgrade checks, but a small set of tests to illustrate the behavior.

### CODE REVIEW

File: tools/pom.xml
Comment (nastra) [inline tools/pom.xml:37]: putting this under `tools` seems rather weird. maybe just put this directly under root?

File: tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/JerseyServer.java
Comment (nastra) [inline tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/JerseyServer.java:77]: just thinking whether it would make sense to make it clearer that these classes are `new` classes that aren't available in the old version (whatever old means)

Comment (nastra) [reply tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/JerseyServer.java:77]: I'm thinking about separating those `new` classes out, something like:

```
private List<Class<?>> classesSinceX() {
    return Arrays.asList(withClass("org.projectnessie.services.rest.RestDiffResource"),withClass("org.projectnessie.services.rest.RestRefLogResource"), ...);
  }
```

and then doing something like `classesSinceX().forEach(config::register);`

File: tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/DatabaseAdapters.java
Comment (nastra) [inline tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/DatabaseAdapters.java:130]: should this be GenericContentVariantSupplier<>(...)?

File: tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/ServerConfigExtension.java
Comment (nastra) [inline tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/ServerConfigExtension.java:26]: seems like a1:1 copy/paste from the existing ServerConfigExtension :(

File: tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/AuthorizerExtension.java
Comment (nastra) [inline tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/AuthorizerExtension.java:30]: 1:1 copy/paste from the existing AuthorizerExtension :(

File: tools/compatibility/common/src/main/java/org/projectnessie/tools/compatibility/internal/TemporaryDirectory.java
Comment (nastra) [inline tools/compatibility/common/src/main/java/org/projectnessie/tools/compatibility/internal/TemporaryDirectory.java:32]: can you elaborate why this is needed vs Junit5's TempDir? I think I know why but would be good to have some comments

File: tools/compatibility/compatibility-tests/src/test/java/org/projectnessie/tools/compatibility/tests/AbstractCompatibilityTests.java
Comment (nastra) [inline tools/compatibility/compatibility-tests/src/test/java/org/projectnessie/tools/compatibility/tests/AbstractCompatibilityTests.java:87]: I'm assuming we'll eventually add many more tests here?

File: tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/ServerConfigExtension.java
Comment (snazy) [reply tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/ServerConfigExtension.java:26]: yea - it just cannot use the other classes due to class loader isolation constraints :(

File: tools/compatibility/compatibility-tests/src/test/java/org/projectnessie/tools/compatibility/tests/AbstractCompatibilityTests.java
Comment (snazy) [reply tools/compatibility/compatibility-tests/src/test/java/org/projectnessie/tools/compatibility/tests/AbstractCompatibilityTests.java:87]: yes (last sentence in the PR description)

File: tools/pom.xml
Comment (snazy) [reply tools/pom.xml:37]: +1

File: tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/JerseyServer.java
Comment (snazy) [reply tools/compatibility/jersey/src/main/java/org/projectnessie/tools/compatibility/jersey/JerseyServer.java:77]: Not a fan of `classesSinceX` (and X being a version number) - that would make the code here very very version specific (it's already a PITA that that module's needed due to class loader isolation requirements).

The alternative to this module would have been a lot of reflection based code - that's an even more a PITA.

**Comment by snazy** [comment]
Pushed 3 more commits - one is the move to the top-level directory

**Review by nastra — APPROVED** [review]
I think it's ok to merge now and do follow-ups if we encounter issues. Having this merged and running CI is more valuable than holding off and making things perfect

**Pull request merged**
